_drafts/partially-ordering-time.md

```diff
@@ -0,0 +1,253 @@
+synchronised”), but if you can prove that a program on your processor is data-race free,
+you can _provide_ sequential consistency to that program as your hardware or software
+memory model (“SQ-DRF”, “sequential consistency for data-race free programs” [6]).
+
+…You can see how this could get confusing.
+
+The obvious solution is, of course, to make things more confusing!
+
+### The Burckhardt Paper
+
+The Burckhardt Paper. The _Burckhardt paper_.
+
+Dr. Sebastian Burckhardt has [been trying](#urls) to formalise the specification of
+consistency models for some time, in the distributed systems world. In [7], Appendix D.4,
+their team lays out an analysis of the C++ weak memory model in a slight specialisation of their framework.
+
+This gives us a convenient way to discuss the similarities and differences between the two
+fields. In particular, the C++ memory model _approximates_ to per-object causal
+consistency, with a few interesting differences.
+
+We'll study them axiom by axiom, but it's first important to note that where causal
+consistency is defined in (Burckhardt's) terms of a “visibility” and an “arbitration”
+orderings, they define weak memory axioms in terms of a “reads-from” and “modification
+order” instead, correspondingly.
+
+“Visibility” in the distributed systems model is a partial order of potential causality;
+it tracks what events (on possibly other replicas) are visible to what other events. It
+can contain read→read, read→write, write→read, and write→write edges, across multiple
+objects. “Reads-from” is much more specialised; it only ever contains write→read edges on
+the same object, and only ever contains one edge for a given read event.
+
+This corresponds to the fact that there only is a single actual memory cell being written
+to in a shared memory processor, for any given object, even if threads may access it at
+different points in causality. In a distributed system, a logical object can be written to
+at many separate locations (“replicas”).
+
+“Arbitration” is a total order that tracks how the distributed system, when asked to make
+a choice, will adjudicate which event happened before which other event. It's distinct
+from “modification order” in that the latter is, again, per-object, and only adjudicates
+writes.
+
+This is, again, a specialisation borne from the fact that the weak memory specification
+only makes strong guarantees per-object.
+
+Okay. With these under our belt, we can discuss each individual property and how it
+differs between distributed systems' per-object causal consistency and the C++11 weak
+memory model.
+
+#### EVENTUAL
+
+For any particular event, there cannot be infinitely many events that do not have
+visibility of it. To make sense under weak memory, this is slightly more complicated: you
+have to state that for any particular _write_, there cannot be infinitely many reads that
+don't read from it or read from an earlier write (by modification order).
```

Comment:
If we call the _write_, write(1), then if some read, read(1) reads from an earlier write, write(0), then does write(1) or write(0) need to know about each other? does read(1) need to know that write(0) was earlier than write(1)? If not, I feel like write(1) can effectively have never existed, which seems much weaker than the distributed system equivalent.

Reply:
If `write(0)` is earlier in modification order than `write(1)`, but all future reads read from `write(0)`, then this is a violation of this eventuality property, yes. Assuming I understand your question, `read(1)` _can't_ read from `write(0)`—the return values don't match.

Reply:
I see, I think the phrasing just needs to be a bit stronger:

> ... there cannot be infinitely many reads that: 
>
> - don't read from the _write_
> - read from an earlier write

When I read it before I grouped it incorrectly, but no matter how I look at it there's only one reasonable way to interpret. But I think it just could read a little more efficiently with a slightly different presentation